cpp/src/sort/rank.cu

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/sorting.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/sequence.h>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+template <bool has_nulls, typename ReturnType = bool>
+struct unique_comparator {
+  unique_comparator(table_device_view device_table,
+                    size_type const *sorted_order)
+      : comp(device_table, device_table, true), perm(sorted_order) {}
+  __device__ ReturnType operator()(size_type index) const noexcept {
+    return index == 0 || not comp(perm[index], perm[index - 1]);
+  };
+
+private:
+  row_equality_comparator<has_nulls> comp;
+  size_type const *perm;
+};
+
+std::unique_ptr<table> rank(table_view const &input, rank_method method,
+                            order column_order, include_nulls _include_nulls,
+                            null_order null_precedence,
+                            bool percentage,
+                            rmm::mr::device_memory_resource *mr,
+                            cudaStream_t stream = 0) {
+  auto const size = input.num_rows();
+
+  std::vector<std::unique_ptr<column>> rank_columns;
+  for (auto const &input_col : input) {
+    std::unique_ptr<column> sorted_order =
+        (method == rank_method::FIRST)
+            ? detail::stable_sorted_order(table_view{{input_col}},
+                                          {column_order}, {null_precedence}, mr,
+                                          stream)
+            : detail::sorted_order(table_view{{input_col}}, {column_order},
+                                   {null_precedence}, mr, stream);
+    column_view sorted_order_view = sorted_order->view();
+
+    // na_option=keep assign NA to NA values
+    if (_include_nulls == include_nulls::NO)
+      rank_columns.push_back(make_numeric_column(
+          data_type(FLOAT64), size, copy_bitmask(input_col, stream, mr),
```

Comment:
Why is this float64?

Reply:
Pandas rank() returns `float64`  type.

Reply:
ewww

Reply:
to handle `na_option="keep"` , rank() returns `nan` for all `nan`s.

Reply:
Isn't this the kind of place where we should do the right thing and let cuDF handle pandas compatibility with `astype`? @kkraus14

Reply:
Strongly agree with @devavret. I do not think we should inherit Pandas' eccentricities that result from the fact that it does not yet properly support null values in non-floating point types.

Reply:
updated: `FLOAT64` for percentage=True and AVERAGE, `INT32` for rest.
used `astype(np.float64)` in pandas.